### PR TITLE
JupyterHub Whitelist

### PR DESCRIPTION
**Pull request by mitkotak**
Add whitelist hosts in `configs/whitelist.json`

### CODE REVIEW

**Comment by alexandermichels** [comment]
Don't have the expertise in the Core and wasn't able to test it, but the code looks simple and straightforward. My one concern would be: how difficult would it be to debug if you weren't in the whitelist (i.e. can I quickly check what the whitelist is to see if my hostname is in there)? Would there be any security vulnerability for making the whitelist available publically (for example as an API route)? Off the top of my head, it wouldn't matter but I'd be curious to hear if that's true.

**Comment by mitkotak** [comment]
Right now there is no way for someone to check whether they are in the whitelist or not. If we expose the whitelist then they can easily change the ``cybergist.jupyterHost`` variable to match one of the names in the whitelist since we do offer a manual login option where they get to choose the host. The best we can do right now is to raise some kind of error in core which then gets communicated through the SDK.

**Comment by alexandermichels** [comment]
Even if they changed their hostname to match a host within the whitelist, the Core will then try to authenticate against the host they specified and since they don't have a valid token, will fail, correct? Or am I missing something?

If tricking our authentication is as simple as putting in a hostname that is in our whitelist, we are screwed because everyone will know that cybergisx is in the whitelist.

**Comment by mitkotak** [comment]
Yup you are right my bad. Yeah even if they know the host, they won't know the token for the host which will be caught by the decode logic. Yeah in that case, I don't see any issues in broadcasting the whitelist. I can just print it out while raising the authentication error. @zimo-xiao What do you think ?

**Comment by alexandermichels** [comment]
Sounds good. Definitely want to get Zimo's opinion first though. I just imagine someone could get stuck spending hours struggling to figure out why they can't interact with their Core deployment lol.

As a follow-up: how often does the whitelist "refresh" from the file? Not sure how typescript importing from JSONs works so I'm curious if I edit the whitelist JSON will that change reflect relatively quickly or will I need to redeploy the code for it to update?

File: configs/whitelist.example.json [hunk 1]
```diff
@@ -0,0 +1,3 @@
+{
+    "whitelist": ["jupyterhub:443"]
```
Comment (zimo-xiao) [inline configs/whitelist.example.json]: Could we add this config into the config.example.json? Add a field like `jupyter_whitelist: []`

File: src/JupyterHub.ts [hunk 2]
```diff
@@ -15,6 +16,16 @@ class JupyterHub {
         var t = this._decodeToken(token)
         const protocols = ['https', 'http']
         var user = undefined
+        var hosts =  whitelistConfig.whitelist
```
Comment (zimo-xiao) [inline src/JupyterHub.ts]: What if the user don't have this config? You might need to add default value like `whitelistConfig.whitelist ?? []`

Comment (mitkotak) [reply src/JupyterHub.ts]: Done !

File: configs/whitelist.example.json (hunk 1 above)
Comment (mitkotak) [reply configs/whitelist.example.json]: Done !

**Comment by mitkotak** [comment]
> Sounds good. Definitely want to get Zimo's opinion first though. I just imagine someone could get stuck spending hours struggling to figure out why they can't interact with their Core deployment lol.
> 
> As a follow-up: how often does the whitelist "refresh" from the file? Not sure how typescript importing from JSONs works so I'm curious if I edit the whitelist JSON will that change reflect relatively quickly or will I need to redeploy the code for it to update?

Sorry for the late reply. I tested it out and looks like we would have to restart the container to see the new whitelist.

**Comment by mitkotak** [comment]
@zimo-xiao Can you have a look now ?

**Pull request closed without merging**